<a href="https://colab.research.google.com/github/lacey-conrad/WGCNA-work/blob/main/Gene_coexp_NN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

import tensorflow


from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.preprocessing import LabelEncoder

from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow import keras
from tensorflow.keras.layers import LeakyReLU

RuntimeError: module compiled against API version 0xf but this version of numpy is 0xe

RuntimeError: module compiled against API version 0xf but this version of numpy is 0xe

ImportError: numpy.core._multiarray_umath failed to import

ImportError: numpy.core.umath failed to import

TypeError: Unable to convert function return value to a Python type! The signature was
	() -> handle

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

# 1. Reading in the dataset:

In [3]:
df = pd.read_table('thermo_capacity_counts.txt')

/var/folders/27/px9hdzpj7vg6cdlnfkj8j2t40000gn/T/ipykernel_50667/1722180382.py:1: DtypeWarning: Columns (5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_table('thermo_capacity_counts.txt')


In [4]:
df_gastroc = pd.read_csv('gastroc_raw_counts.csv')

In [5]:
#from google.colab import drive
#drive.mount('/content/drive')

In [6]:
#/content/thermo_capacity_counts.txt

In [7]:
#df_copy = df.copy()

In [8]:
df_gastroc.head(50)

Unnamed: 0  LN.F1.113.2be  LN.F1.113.aa  LN.F1.113.ae  LN.F1.113.ag  \
0   LOC121821266              0             0             0             0   
1   LOC102911281              0             0             0             0   
2   LOC102926907              0             0             0             0   
3   LOC121821617              0             0             0             1   
4   LOC102925994              3             4             3             4   
5   LOC102925679              0             0             0             0   
6   LOC121821541              3             1             1             0   
7   LOC102925362              3             0             2             0   
8   LOC121821593              0             0             0             0   
9   LOC102924750              0             0             0             0   
10  LOC102919686              0             0             0             0   
11  LOC102919997              0             0             0             0   
12  LOC102920311              0             0             0             0   
13  LOC102918165              0             0             0             0   
14  LOC102927417              7             4            10             4   
15  LOC102921144              0             0             0             2   
16  LOC102927115              0             0             0             0   
17  LOC121820855              0             0             0             0   
18  LOC102911923              0             0             0             0   
19  LOC121821653              0             1             0             0   
20  LOC102918878              0             2             1             1   
21  LOC102926911              0             0             0             0   
22  LOC121827067              0             0             0             0   
23  LOC102918977              0             0             0             0   
24  LOC102922252              0             0             0             0   
25  LOC102928593              0             0             0             0   
26  LOC102928909              0             0             0             0   
27  LOC107400139              0             0             0             0   
28  LOC102903402              0             0             0             0   
29  LOC102903704              0             0             0             0   
30  LOC102904335              0             0             0             0   
31  LOC102904643              0             0             0             0   
32  LOC102903073             10            12             5             6   
33  LOC102907096              0             0             0             0   
34  LOC121824963              0             0             0             0   
35  LOC121825743              0             0             0             0   
36  LOC102907711              0             0             0             0   
37  LOC102908312              0             0             0             0   
38  LOC102908633              0             0             0             0   
39  LOC102903383              0             0             0             0   
40  LOC121822775              0             0             0             0   
41  LOC102909563              0             0             0             0   
42  LOC102904318             27            35            21            25   
43  LOC121822122              3             4             1             1   
44  LOC121822734              0             1             0             1   
45  LOC102903682              4             5             4            11   
46  LOC102911786              0             0             0             0   
47  LOC102912105              6            10             3             7   
48       Ppp2r1a            336           264           175           298   
49  LOC121822845              0             0             0             0   

    LN.F1.113.ak  LN.F1.113.am  LN.F1.113.b  LN.F1.113.h  LN.F1.113.p  ...  \


In [9]:
df_gastroc['Row_Mean'] = df_gastroc.mean(axis=1)

/var/folders/27/px9hdzpj7vg6cdlnfkj8j2t40000gn/T/ipykernel_50667/2138856592.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_gastroc['Row_Mean'] = df_gastroc.mean(axis=1)


In [10]:
df_gastroc.head()

Unnamed: 0  LN.F1.113.2be  LN.F1.113.aa  LN.F1.113.ae  LN.F1.113.ag  \
0  LOC121821266              0             0             0             0   
1  LOC102911281              0             0             0             0   
2  LOC102926907              0             0             0             0   
3  LOC121821617              0             0             0             1   
4  LOC102925994              3             4             3             4   

   LN.F1.113.ak  LN.F1.113.am  LN.F1.113.b  LN.F1.113.h  LN.F1.113.p  ...  \
0             0             0            0            0            0  ...   
1             0             0            0            0            0  ...   
2             0             0            0            0            0  ...   
3             0             0            0            0            0  ...   
4             7             7            9            2            6  ...   

   ME.F1.145.d  ME.F1.145.g  ME.F1.145.i  ME.F1.145.w  ME.F1.146.aa  \
0            0            0            0            0             0   
1            0            0            0            0             0   
2            0            0            0            0             0   
3            1            0            1            0             0   
4            9            5            6           11             2   

   ME.F1.146.ab  ME.F1.146.ac  ME.F1.146.be  ME.F1.146.bf  Row_Mean  
0             0             0             0             0  0.000000  
1             0             0             0             0  0.000000  
2             0             0             0             0  0.000000  
3             0             1             1             0  0.404255  
4             2             4             2             5  4.319149  

[5 rows x 49 columns]

In [11]:
df_gastroc = df_gastroc[df_gastroc.Row_Mean >= 10]

In [12]:
df_gastroc.head()

Unnamed: 0  LN.F1.113.2be  LN.F1.113.aa  LN.F1.113.ae  LN.F1.113.ag  \
32  LOC102903073             10            12             5             6   
42  LOC102904318             27            35            21            25   
48       Ppp2r1a            336           264           175           298   
55  LOC102906476            384           340           407           446   
60  LOC102922856              6            13            21            21   

    LN.F1.113.ak  LN.F1.113.am  LN.F1.113.b  LN.F1.113.h  LN.F1.113.p  ...  \
32             7             8            9           12           11  ...   
42            25            24           37           23           25  ...   
48           425           288          230          326          262  ...   
55           526           577          343          500          447  ...   
60            17            15           18           16           18  ...   

    ME.F1.145.d  ME.F1.145.g  ME.F1.145.i  ME.F1.145.w  ME.F1.146.aa  \
32           14           11           23           20             6   
42           20           27           37           35            18   
48          182          188          426          311           290   
55          364          332          642          588           350   
60            8           18           25           32            16   

    ME.F1.146.ab  ME.F1.146.ac  ME.F1.146.be  ME.F1.146.bf    Row_Mean  
32             8             6            10            19   10.680851  
42            18            19            13            22   22.787234  
48           324           266           204           306  292.148936  
55           371           331           345           461  400.617021  
60             8            11            14            24   16.468085  

[5 rows x 49 columns]

In [ ]:
df_gastroc = df_gastroc.T

In [ ]:
df_gastroc.columns = df_gastroc.iloc[0]

In [ ]:
df_gastroc = df_gastroc.iloc[1: , :]

In [ ]:
df_gastroc.head()

In [ ]:
df_gastroc = df_gastroc.reset_index()

In [ ]:
df_gastroc.head()

In [ ]:
df_gastroc = df_gastroc.rename(columns={'index': 'Treatment'})

In [ ]:
df_gastroc.head()

In [ ]:
df_gastroc['Treatment'] = df_gastroc['Treatment'].str.split('.').str[0]

In [ ]:
df_gastroc.head()

In [ ]:
cleanup_nums = {"Treatment": {"LN": 0, "ME": 1}}

In [ ]:
df_gastroc = df_gastroc.replace(cleanup_nums)
df_gastroc.head()

In [ ]:
X = df_gastroc.drop(labels=['Treatment'], axis = 1)
y = df_gastroc.Treatment

In [ ]:
genes = list(X.columns)

In [ ]:
#genes

In [ ]:
#df_gastroc_mat = X.to_numpy()

In [ ]:
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)

In [ ]:
X = pd.DataFrame(X)

In [ ]:
X.head()

In [ ]:
df_gastroc_mat = X.to_numpy()

In [ ]:
X = df_gastroc_mat.astype(int)

In [ ]:
X_norm_genes = pd.DataFrame(data = df_gastroc_mat, columns = genes)

In [ ]:
X_norm_genes.head()

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_norm_genes, y, test_size=0.20, random_state = 42)

In [ ]:
#X.head()

In [ ]:
# Model#1
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 10

In [ ]:
# Initializing the ANN
model = Sequential()

# Adding the input layers and hidden layers:
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))

# Adding the output layer:
model.add(Dense(1, activation = 'sigmoid'))

# Compiling the ANN:
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])



In [ ]:
# Training the ANN:
#model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 10, verbose=1)

In [ ]:
# Making predictions:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
# Predictions of the response variable (whether or not the customer will exit the bank)
y_pred

In [ ]:
# Evaluating the model:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#2 - more nodes
# Input layer: 50 nodes
# Hidden layer 1: 500 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 10

In [ ]:
model = Sequential()
model.add(Dense(50, activation='relu'))
model.add(Dense(500, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 10, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#3 - larger batch size
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 500

In [ ]:
model = Sequential()
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 500, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#4 - More epochs
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 100
# Batch size: 10

In [ ]:
model = Sequential()
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 100, batch_size = 10, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

# Insights derived from altering model hyperparameters:

- A smaller batch size produced more accurate results, but took longer to run.

- Adding more nodes in the hidden layers did not seem to improve the accuracy of the model.

- Increasing the number of epochs decreased the accuracy.

- My best model was number one:
 
 Model#1
 
 Input layer: 32 nodes
 
 Hidden layer 1: 64 nodes
 
 Optimizer: Adam
 
 Loss: Binary Crossentropy
 
 Metrics: Accuracy
 
 Epochs: 10
 
 Batch size: 10
 
True negatives: 1514 (Those who my model predicted would not exit, who did not actually exit)

False negatives: 79 (Those who my model predicted would exit, but who did not actually exit)

False positive: 200 (Those who my model predicted would not exit, but who did actually exit)

True positives: 207 (Those who my model predicted would exit the bank, who did actually exit the bank)

 accuracy of model: 86%